File: app/src/main/java/com/example/hiltproject/rxJavaLearning/MainActivity.java
package com.example.hiltproject.rxJavaLearning;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.widget.ProgressBar;

import com.example.hiltproject.R;
import com.example.hiltproject.WeatherModel;
import com.example.hiltproject.network.RetrofitService;
import com.example.hiltproject.repo.MainRepo;
import com.example.hiltproject.roomDatabase.WeatherDao;
import com.example.hiltproject.roomDatabase.WeatherDatabase;
import com.example.hiltproject.ui.WeatherAdapter;

import java.util.ArrayList;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.functions.Consumer;
import io.reactivex.rxjava3.schedulers.Schedulers;
import kotlinx.coroutines.GlobalScope;
import rx.Subscription;
import rx.functions.Action1;
import rx.subscriptions.CompositeSubscription;

public class MainActivity extends AppCompatActivity implements WeatherAdapter.ClickListener {

    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    WeatherAdapter weatherAdapter ;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main2);
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        weatherAdapter = new WeatherAdapter(this);

        recyclerView.setAdapter(weatherAdapter);


        compositeDisposable.add(ApiClient
                .getInstance()
                .getWeatherList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(detail -> {
                    if (detail != null) {
                        weatherAdapter.updateList(detail.getDataseries());}

                }, throwable -> {

                    throwable.printStackTrace();
                    Log.d("throwable_m" , throwable.getLocalizedMessage());
                }));
    }

    @Override
    public void onItemClickListener(int position) {

    }
}

File: app/src/main/java/com/example/hiltproject/ui/MainActivity.kt
package com.example.hiltproject.ui

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.hiltproject.WeatherDataModel
import com.example.hiltproject.databinding.ActivityMainBinding
import com.example.hiltproject.repo.MainRepo
import com.example.hiltproject.roomDatabase.WeatherDao
import com.example.hiltproject.roomDatabase.WeatherDatabase
import com.example.hiltproject.viewModel.MainActivityViewModel


class MainActivity : AppCompatActivity(){
    lateinit private var repository: MainRepo
    lateinit var viewModel: MainActivityViewModel
    lateinit var weatherAdapter : WeatherAdapter
    private var binding: ActivityMainBinding?=null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding?.root
        setContentView(view)
        setupRecyclerView()
        val  dao = WeatherDatabase.invoke(application).getWeatherDao()
        setupViewModel(dao)
        observeData(dao)
    }

    private fun setupViewModel(dao: WeatherDao) {
        repository = MainRepo(dao)
        viewModel = ViewModelProviders.of(this,
            MainActivityViewModel.ViewModelFactory(repository)).get(MainActivityViewModel::class.java)
        binding?.progressCircular!!.visibility= View.VISIBLE

        repository .getDetail(applicationContext)
    }

    private fun observeData(dao: WeatherDao) {
        binding?.progressCircular!!.visibility= View.GONE

        dao.getAllData().observe(this) {
            if (it.isNullOrEmpty()) {
               repository .getDetail(applicationContext)
            }else{
                binding?.progressCircular!!.visibility = View.GONE
                weatherAdapter.updateList(it as ArrayList<WeatherDataModel> /* = java.util.ArrayList<com.example.hiltproject.WeatherDataModel> */)

            }
        }

    }

    private fun setupRecyclerView() {
        binding?.recyclerView?.apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            weatherAdapter=WeatherAdapter(object : WeatherAdapter.ClickListener{
                override fun onItemClickListener(position: Int) {
                    val intent= Intent(this@MainActivity,WeatherDetailActivity::class.java)
                    intent.putExtra("POSITION",position)
                    startActivity(intent)
                }
            })
            adapter=weatherAdapter
        }

    }

}